src/Microsoft.AspNetCore.Mvc.TagHelpers/Cache/CacheTagKey.cs

```diff
@@ -0,0 +1,319 @@
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers.Cache
+{
+    /// <summary>
+    /// An instance of <see cref="CacheTagKey"/> represents the state of <see cref="CacheTagHelper"/>
+    /// or <see cref="DistributedCacheTagHelper"/> keys.
+    /// </summary>
+    public class CacheTagKey : IEquatable<CacheTagKey>
+    {
+        private static readonly char[] AttributeSeparator = new[] { ',' };
+
+        private const string CacheKeyTokenSeparator = "||";
+        private const string VaryByName = "VaryBy";
+        private const string VaryByHeaderName = "VaryByHeader";
+        private const string VaryByQueryName = "VaryByQuery";
+        private const string VaryByRouteName = "VaryByRoute";
+        private const string VaryByCookieName = "VaryByCookie";
+        private const string VaryByUserName = "VaryByUser";
+
+        private string _key;
+        private string _prefix;
+        private string _varyBy;
+        private DateTimeOffset? _expiresOn;
+        private TimeSpan? _expiresAfter;
+        private TimeSpan? _expiresSliding;
+        private IList<KeyValuePair<string, string>> _headers;
+        private IList<KeyValuePair<string, string>> _queries;
+        private IList<KeyValuePair<string, string>> _routeValues;
+        private IList<KeyValuePair<string, string>> _cookies;
+        private bool _varyByUser;
+        private string _username;
+
+        private string _generatedKey;
+        private int? _hashcode;
+        
+        
+        /// <summary>
+        /// Creates an instance of <see cref="CacheTagKey"/> for a specific <see cref="CacheTagHelper"/>.
+        /// </summary>
+        /// <param name="tagHelper">The <see cref="CacheTagHelper"/>.</param>
+        /// <param name="context">The <see cref="TagHelperContext"/>.</param>
+        /// <returns>A new <see cref="CacheTagKey"/>.</returns>
+        public CacheTagKey(CacheTagHelper tagHelper, TagHelperContext context)
+            : this((CacheTagHelperBase)tagHelper, context)
+        {
+            _key = context.UniqueId;
+            _prefix = nameof(CacheTagHelper);
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="CacheTagKey"/> for a specific <see cref="DistributedCacheTagHelper"/>.
+        /// </summary>
+        /// <param name="tagHelper">The <see cref="DistributedCacheTagHelper"/>.</param>
+        /// <param name="context">The <see cref="TagHelperContext"/>.</param>
+        /// <returns>A new <see cref="CacheTagKey"/>.</returns>
+        public CacheTagKey(DistributedCacheTagHelper tagHelper, TagHelperContext context)
+            : this((CacheTagHelperBase)tagHelper, context)
+        {
+            _key = tagHelper.Name;
+            _prefix = nameof(DistributedCacheTagHelper);
+        }
+
+        private CacheTagKey(CacheTagHelperBase tagHelper, TagHelperContext context)
+        {
+            var httpContext = tagHelper.ViewContext.HttpContext;
```

Comment:
Don't need these guys unless you're varying by user. Move them down to that block.